### DISCUSSION TITLE
Distinguish between an INSERT from a trigger vs supabase client

### DISCUSSION THREAD

**Discussion by jasonho-lynx** [supabase/supabase#1290] [Questions] [answered]
Hi, is there a way to distinguish between an INSERT operation from a database trigger, vs an INSERT from the supabase client? 

In my case, I have a `clean_input()` trigger for `test_table`, such that anytime before an INSERT is done, it will clean up the entries before inserting it into the database (e.g. setting `last_modified_time` to `CURRENT_TIMESTAMP`, `last_user` etc).

This is to clean up inputs from the client in case inaccurate values are set e.g. `await supabase.from('test_table').insert([{ last_modified_time: <some time in the future>}]);`.
 
At the same time, I also have another trigger function e.g. `insert_into_test_table()`, which is invoked elsewhere in Postgres and inserts into the same `test_table`. Would it be possible for `clean_input()` to ignore an INSERT from `insert_into_test_table()` i.e. ignore 'server-side' INSERTs? 

Quite new to Supabase and Postgres so also not sure if this is the best approach, happy to hear better solutions if you have any. Thank you!

**Comment by steve-chavez** [comment, marked as answer]
Hey @jasonho-lynx,

Since `clean_input` expects input from http clients(through supabase-js), you could query an http attribute to differentiate between a client-side and server-side insert.

In particular, the `request.method` setting might be used like this;

```sql
create or replace function clean_client_input() returns trigger as $$
declare
  is_post bool := current_setting('request.method', true) = 'POST';
begin
  if is_post then
    -- perform cleaning
    return new;
  else
    return new; -- do nothing
  end if;
end;
$$ language plpgsql;
```

http settings are documented [here](https://postgrest.org/en/stable/api.html#http-logic). Obtaining the client jwt claims(perhaps through the already available `auth.uid()`) would also work, but obtaining the http method might be more fitting here.

**Reply by jasonho-lynx** [reply]
This looks great, thanks a lot @steve-chavez! :)

I also happened to find a [post](https://dba.stackexchange.com/questions/103402/how-to-prevent-a-postgresql-trigger-from-being-fired-by-another-trigger) suggesting the use of `pg_trigger_depth() < 1` as a condition on the `clean_input` trigger to achieve a similar goal. Would you recommend one over the other?

Edit: gotta say I still prefer your solution since it's more explicit and intentional

**Reply by jasonho-lynx** [reply]
FYI for anyone else bumping onto this thread, if you'd also like to clean `UPDATE` queries, they use `PATCH` methods so it could be something like:

```
DECLARE 
    is_post_or_patch := CURRENT_SETTING('request.method`, true) IN ('POST', 'PATCH');
...
```

**Reply by steve-chavez** [reply]
> Would you recommend one over the other?
> Edit: gotta say I still prefer your solution since it's more explicit and intentional

Yeah, it's definitely more intentional. Perhaps you could even rename the trigger function to `clean_api_client_input` or `clean_http_client_input` so any other dev/dba that sees the trigger can immediately know what it's doing.

Regarding `pg_trigger_depth`, I agree with this so answer: [it's always bad to make behavior dependent on pg_trigger_depth()](https://dba.stackexchange.com/a/163173/158470). It's just too hacky IMHO, can lead to surprising behavior.

Also, another option is to make the insert through a security definer function(no GRANT INSERT for clients), that would avoid a trigger.

**Reply by jasonho-lynx** [reply]
Sorry this will be a bit lengthy, but I'm trying to summarise the options available and their implications. Please correct me if I'm wrong:

## Option 1
#### `clean_http_client_input()` trigger functions with `SECURITY DEFINER`
- No extra configuration for permissions
- If this trigger function's effects cascade and set off other trigger functions, the latter triggers will have `current_setting('request.method', true)` set to `NULL`, so it will go  `clean_http_client_input_t1()` -> `insert_into_t2_after_t1_insert()`. 

(EDIT: after testing things a bit more it seems like this assumption is **wrong**, at least sometimes. Am seeing this to be true in some cases and not in others)

(EDIT 2: was able to work around it by doing a `PERFORM SET_CONFIG('request.method', NULL, TRUE);` right before the end of the trigger function. Not sure if it's a bad idea.)

## Option 2
#### `clean_http_client_input()` trigger functions with `SECURITY INVOKER` default
- Need to `GRANT EXECUTE` on the `auth.uid()` function to `authenticated` user if it's used
- `authenticated` user must also have `SELECT` permissions on any other tables referenced in the trigger function
- If this trigger function's effects cascade and set off other trigger functions (say for logging or other automatic events), the latter triggers will also be treated as client-side `INSERTs` despite being fired off server-side (e.g. `clean_http_client_input_t1()` -> `insert_into_t2_after_t1_insert()` -> `clean_http_client_input_t2()`)

## Option 3
#### Remove `INSERT` for clients (i.e. `REVOKE INSERT ON DATABASE my_database FROM 'authenticated'`), and make all inserts via `SECURITY DEFINER` functions (not trigger functions)
- `GRANT EXECUTE` on these functions to `authenticated` user
- Every supabase function will need a corresponding Postgres function and each will be called as follows: `const myFunction = () => await supabase.rpc('myFunction')`

---
Based on these, my preference is for option 1. Does that make sense?

**Reply by steve-chavez** [reply]
> Option 3

This is the one that I referred to above, sorry if I wasn't clear about using this through `supabase.rpc`.

> Based on these, my preference is for option 1. Does that make sense?

In general I think you should go with option 2, and have special care for what tables you grant access to clients.

> Option 2
> Need to GRANT EXECUTE on the auth.uid() function to authenticated user if it's used

This one will likely be patched by us, authenticated will have access to the `auth` schema functions by default.

> the latter triggers will also be treated as client-side INSERTs despite being fired off server-side

Oh, is this true even for security definer triggers? Say if you fire a security invoker trigger that cascades to another trigger that has security definer, will the latter have its privileges restricted?

> Option 1

> If this trigger function's effects cascade and set off other trigger functions, the latter triggers will have current_setting('request.method', true) set to NULL

No, the `request.method` will have the `POST` value for the entire transaction, even for the subsequent triggers.

What happens on each request, is essentially:

```sql
begin;
set local role <the_jwt_role>.;
set local "request.method" = 'POST';
-- ... other settings
insert into .... -- the insert query
-- ...
end; -- end of transaction
```

**Reply by jasonho-lynx** [reply]
> This one will likely be patched by us, authenticated will have access to the auth schema functions by default.

Got it - are there additional nuances to consider or is it a matter of adding these lines 
```
GRANT USAGE ON SCHEMA auth TO authenticated;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA auth TO authenticated;
```
to https://github.com/supabase/supabase/blob/master/docker/postgres/auth-schema.sql? If it's any help at all, I could try to create a PR for that.

> Oh, is this true even for security definer triggers? Say if you fire a security invoker trigger that cascades to another trigger that has security definer, will the latter have its privileges restricted?

Ah no the latter won't have its privileges restricted. I wasn't being clear but when I mentioned 

> the latter triggers will also be treated as client-side INSERTs despite being fired off server-side

I was actually referring to `current_setting('request.method', true)` still being equal to `POST` even for latter triggers, which you also confirmed later in your comment.

I'd actually edited my comment above to say the same, as well as add a workaround, but it probably came too late.

In that case, I'll go with something like this:

- Use `SECURITY INVOKER` for the trigger functions as much as possible
- Each `clean_http_client_input()` trigger function will be basically what you provided in the very first comment, with a small addition to reset `request.method`  before `return new`:
 
```
create or replace function clean_http_client_input() returns trigger as $$
declare
  is_post bool := current_setting('request.method', true) = 'POST';
begin
  if is_post then
    -- perform cleaning
    ...

    -- reset request.method to mark the end of a client request, so that cascading triggers will not clean input
    reset "request.method";

    return new;
  else
    return new; -- do nothing
  end if;
end;
$$ language plpgsql;
```

Again, this was very helpful, thanks! Let me know if there's anything I can do for the patching bit - don't know much but will try to help where I can.

**Reply by steve-chavez** [reply]
>  If it's any help at all, I could try to create a PR for that.

Sure. That would be great! 

> In that case, I'll go with something like this:
>  -- reset request.method to mark the end of a client request, so that cascading triggers will not clean input
> reset "request.method";

Hm, but the other triggers need to fire up whether is a client-side or server-side insert right? Why the need to reset the `request.method` if only the cleaning input trigger uses it?

Btw, perhaps it'd even be more clear to use the `WHEN` clause on the trigger:

```sql
create trigger clean_http_input 
before insert on my_table
for each row
when (current_setting('request.method', true) = 'POST')
execute function clean_http_input();
```

**Reply by jasonho-lynx** [reply]
PR here: https://github.com/supabase/supabase/pull/1377 - hope it's right!

> Btw, perhaps it'd even be more clear to use the WHEN clause on the trigger:

Oo nice, yes I'm going to use this. Thanks!

> Hm, but the other triggers need to fire up whether is a client-side or server-side insert right?

Not all of them - I want server-side inserts to be ignored by any subsequent cleaning input triggers. A contrived but similar example would be something like this:

- There are 2 tables that a client can directly insert into: `folders` and `files`.
    - Thus for each table, there will be a `clean_http_input` trigger before inserts
    - Both rely on checking `request.method` to see if it came from the client
- There is also a server-side trigger called `create_first_file` that inserts into `files` after an insert into `folders` (e.g. when a new folder is created, a sample 'my-first-file' is also automatically created)
    - These inserts should be ignored by `clean_http_input_files()`.

Since `request.method` has a `POST` value for the entire transaction by default, if a client inserts into `folders`, the process would be:
```
-- not intended
[POST] clean_http_input_folders() -> [POST] create_first_file() -> [POST] clean_http_input_files()
```
`clean_http_input_files()` detects that there's still a POST value, and cleans the server-side insert from `create_first_file()`, erroneously overwriting some values.

Instead, by resetting `request.method` at the end of every `clean_http_input` trigger, I worry less about the overall transaction context, and whether it has a `request.method` or not. For an insert into any table, if it has a `request.method`, then it's from the client (clean it). If not, it's from the server (leave it).

```
-- intended. clean_http_input_files() does not run
[POST] clean_http_input_folders() -> [''] create_first_file()
```
